or-tools: 9.12 -> 9.14

flokli opened the pull request:
or-tools: 9.12 -> 9.14
    
    Some of the patches we ship(ed) had to be regenerated.
    
    or-tools now uses abseil-cpp_202505, and we need to ensure everything
    uses that, to avoid another version of abseil leaking into the build.
    
    pybind11 needs to be downgraded to 2.13.6 (I vendored the old nixpkgs
    expression for that).
    pybind11 also needs to be patched to fix math_opt's elemental python
    binding.
    or-tools ships this in their repo as a patch to their pybind11_bazel
    dep, including a patch file there. I decided to pop one layer of
    patching, and vendor the patch in here.
    
    Upstream also patches their scipopts-scip dependency, renaming some
    CMake targets, so we need to patch our dependency as well (and restore
    find_package() discovery that broke).

---

    or-tools: move pybind11-protobuf package into or-tools
    
    or-tools is the only consumer of this library, and
    1910563bb3c0a76aae035d578a8c094c5646baa7 is already one instance of the
    toplevel build being broken.
    
    Absorb pybind11-protobuf into or-tools, and manage the pinning there. It
    will also make it much easier to do the upcoming or-tools bump, which
    will also change these version bounds.
    
    pybind11-protobuf is not actually packaged in the version that or-tools
    asks for, but trying to use that version showed it wants abseil from
    2023, which has been removed from nixpkgs.

---

    Revert "abseil-cpp_202505: drop"
    
    This reverts commit 7559006efc5254e8e096084fd8dd349306430acd.
    
    It is needed for the current or-tools release, 9.14:
    
    https://github.com/google/or-tools/blob/93dafb79a0261b5211f4eb17437aeebd7f45e543/MODULE.bazel#L16

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

flokli committed:
Revert "abseil-cpp_202505: drop"

This reverts commit 7559006efc5254e8e096084fd8dd349306430acd.

It is needed for the current or-tools release, 9.14:

https://github.com/google/or-tools/blob/93dafb79a0261b5211f4eb17437aeebd7f45e543/MODULE.bazel#L16

flokli committed:
or-tools: move pybind11-protobuf package into or-tools

or-tools is the only consumer of this library, and
1910563bb3c0a76aae035d578a8c094c5646baa7 is already one instance of the
toplevel build being broken.

Absorb pybind11-protobuf into or-tools, and manage the pinning there. It
will also make it much easier to do the upcoming or-tools bump, which
will also change these version bounds.

pybind11-protobuf is not actually packaged in the version that or-tools
asks for, but trying to use that version showed it wants abseil from
2023, which has been removed from nixpkgs.

flokli committed:
or-tools: 9.12 -> 9.14

Some of the patches we ship(ed) had to be regenerated.

or-tools now uses abseil-cpp_202505, and we need to ensure everything
uses that, to avoid another version of abseil leaking into the build.

pybind11 needs to be downgraded to 2.13.6 (I vendored the old nixpkgs
expression for that).
pybind11 also needs to be patched to fix math_opt's elemental python
binding.
or-tools ships this in their repo as a patch to their pybind11_bazel
dep, including a patch file there. I decided to pop one layer of
patching, and vendor the patch in here.

Upstream also patches their scipopts-scip dependency, renaming some
CMake targets, so we need to patch our dependency as well (and restore
find_package() discovery that broke).

flokli committed:
or-tools: enable tests on aarch64-linux

https://github.com/google/or-tools/issues/4746 got fixed upstream,
apply the fix locally until part of a release.

flokli commented:
The last successful build for openroad on aarch64-linux was in 2023-03-27: https://hydra.nixos.org/build/311798039, so this can be ignored. This smells like an upstream issue, assuming math (and due to this synthesis) behaves the same on all architectures.

I can take a look at the Darwin failures. Can you link the build logs somewhere?

GaetanLepage commented:
> The last successful build for openroad on aarch64-linux was in 2023-03-27: https://hydra.nixos.org/build/311798039, so this can be ignored. This smells like an upstream issue, assuming math (and due to this synthesis) behaves the same on all architectures.
> 
> I can take a look at the Darwin failures. Can you link the build logs somewhere?

Yes, let's forget about the failing ARM build.
The darwin failures are unrelated too.

flokli commented on pkgs/top-level/python-packages.nix:
```diff
@@ -12737,8 +12737,6 @@ self: super: with self; {
   pybind11-abseil = callPackage ../development/python-modules/pybind11-abseil { };
 
-  pybind11-protobuf = callPackage ../development/python-modules/pybind11-protobuf { };
-
```
This is in the python package set, not in toplevel.

GaetanLepage commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 458786`
Commit: `1f3b369dbfc46d4647f4798979843e9af5899cb1`

---
### `x86_64-linux`
<details>
  <summary>:x: 4 packages failed to build:</summary>
  <ul>
    <li>openroad</li>
    <li>or-tools</li>
    <li>or-tools.python (python313Packages.ortools, python313Packages.ortools.python)</li>
    <li>python312Packages.ortools (python312Packages.ortools.python)</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>abseil-cpp_202505</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:x: 4 packages failed to build:</summary>
  <ul>
    <li>openroad</li>
    <li>or-tools</li>
    <li>or-tools.python (python313Packages.ortools, python313Packages.ortools.python)</li>
    <li>python312Packages.ortools (python312Packages.ortools.python)</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>abseil-cpp_202505</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:x: 4 packages failed to build:</summary>
  <ul>
    <li>openroad</li>
    <li>or-tools</li>
    <li>or-tools.python (python313Packages.ortools, python313Packages.ortools.python)</li>
    <li>python312Packages.ortools (python312Packages.ortools.python)</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>abseil-cpp_202505</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:x: 4 packages failed to build:</summary>
  <ul>
    <li>openroad</li>
    <li>or-tools</li>
    <li>or-tools.python (python313Packages.ortools, python313Packages.ortools.python)</li>
    <li>python312Packages.ortools (python312Packages.ortools.python)</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>abseil-cpp_202505</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `x86_64-linux`</summary>
<details>
<summary>or-tools</summary>
<pre>[  SKIPPED ] HighsIncrementalMultiObjectiveTest/IncrementalMultiObjectiveTest.ModifyAuxiliaryObjectiveSenseInMultiObjectiveModel/0
[  FAILED  ] 1 test, listed below:
[  FAILED  ] HighsMessageCallbackTest/MessageCallbackTest.ObjectiveValueAndEndingSubstring/1, where GetParam() = { solver_type: highs, support_message_callback: 1, support_interrupter: 0, integer_variables: 1, ending_substring: &quot;(heuristics)&quot;, solve_parameters: gscip { } gurobi { } glop { } cp_sat { } pdlp { } glpk { } highs { } }

 1 FAILED TEST

648/652 Test  #31: cxx_graph_min_cost_flow_test ...............................................   Passed   55.59 sec
649/652 Test #606: python_python_no_wait_baking_scheduling_sat ................................   Passed   35.63 sec
650/652 Test #173: python_linear_solver_model_builder_test ....................................   Passed   56.86 sec
651/652 Test  #14: cxx_algorithms_radix_sort_test .............................................   Passed   68.06 sec
652/652 Test  #55: cxx_sat_cp_model_presolve_random_test ......................................   Passed  145.55 sec

99% tests passed, 1 tests failed out of 652

Total Test time (real) = 146.13 sec

The following tests FAILED:
	166 - cxx_math_opt_solvers_highs_solver_test (Failed)
Errors while running CTest
make: *** [Makefile:71: test] Error 8</pre>
</details>
<details>
<summary>python312Packages.ortools</summary>
<pre>642/652 Test #160: cxx_math_opt_solvers_gscip_solver_test .....................................   Passed   38.03 sec
643/652 Test #487: python_contrib_magic_sequence_sat ..........................................   Passed   33.27 sec
644/652 Test #606: python_python_no_wait_baking_scheduling_sat ................................   Passed   33.88 sec
645/652 Test #616: python_python_qubo_sat .....................................................   Passed   34.79 sec
646/652 Test #344: cxx_math_opt_time_indexed_scheduling .......................................   Passed   44.90 sec
647/652 Test #510: python_contrib_permutation_flow_shop .......................................   Passed   54.77 sec
648/652 Test  #78: cxx_sat_integer_expr_test ..................................................   Passed   62.51 sec
649/652 Test  #31: cxx_graph_min_cost_flow_test ...............................................   Passed   65.40 sec
650/652 Test #173: python_linear_solver_model_builder_test ....................................   Passed   66.53 sec
651/652 Test  #14: cxx_algorithms_radix_sort_test .............................................   Passed   78.28 sec
652/652 Test  #55: cxx_sat_cp_model_presolve_random_test ......................................   Passed  154.30 sec

99% tests passed, 1 tests failed out of 652

Total Test time (real) = 154.41 sec

The following tests FAILED:
	166 - cxx_math_opt_solvers_highs_solver_test (Failed)
Errors while running CTest
make: *** [Makefile:71: test] Error 8</pre>
</details>
</details>

---
<details>
<summary>Error logs: `aarch64-linux`</summary>
<details>
<summary>or-tools</summary>
<pre>642/652 Test #649: python_tests_issue4 ........................................................   Passed    2.90 sec
643/652 Test #632: python_python_weighted_latency_problem_sat .................................   Passed    8.55 sec
644/652 Test #622: python_python_spread_robots_sat ............................................   Passed   24.86 sec
645/652 Test #606: python_python_no_wait_baking_scheduling_sat ................................   Passed   32.89 sec
646/652 Test  #31: cxx_graph_min_cost_flow_test ...............................................   Passed   69.54 sec
647/652 Test #616: python_python_qubo_sat .....................................................   Passed   33.47 sec
648/652 Test  #14: cxx_algorithms_radix_sort_test .............................................   Passed   77.33 sec
649/652 Test #173: python_linear_solver_model_builder_test ....................................   Passed   76.17 sec
650/652 Test #510: python_contrib_permutation_flow_shop .......................................   Passed   53.12 sec
651/652 Test #351: python_math_opt_time_indexed_scheduling ....................................   Passed   68.66 sec
652/652 Test  #55: cxx_sat_cp_model_presolve_random_test ......................................   Passed  177.53 sec

99% tests passed, 1 tests failed out of 652

Total Test time (real) = 178.64 sec

The following tests FAILED:
        166 - cxx_math_opt_solvers_highs_solver_test (Failed)
Errors while running CTest
make: *** [Makefile:71: test] Error 8</pre>
</details>
<details>
<summary>python312Packages.ortools</summary>
<pre>642/652 Test #613: python_python_prize_collecting_vrp_sat .....................................   Passed   19.77 sec
643/652 Test #623: python_python_steel_mill_slab_sat ..........................................   Passed   10.29 sec
644/652 Test #351: python_math_opt_time_indexed_scheduling ....................................   Passed   47.34 sec
645/652 Test #632: python_python_weighted_latency_problem_sat .................................   Passed    7.37 sec
646/652 Test #622: python_python_spread_robots_sat ............................................   Passed   17.73 sec
647/652 Test #606: python_python_no_wait_baking_scheduling_sat ................................   Passed   32.62 sec
648/652 Test  #14: cxx_algorithms_radix_sort_test .............................................   Passed   80.40 sec
649/652 Test #510: python_contrib_permutation_flow_shop .......................................   Passed   50.74 sec
650/652 Test #173: python_linear_solver_model_builder_test ....................................   Passed   76.20 sec
651/652 Test #616: python_python_qubo_sat .....................................................   Passed   33.51 sec
652/652 Test  #55: cxx_sat_cp_model_presolve_random_test ......................................   Passed  174.46 sec

99% tests passed, 1 tests failed out of 652

Total Test time (real) = 175.61 sec

The following tests FAILED:
        166 - cxx_math_opt_solvers_highs_solver_test (Failed)
Errors while running CTest
make: *** [Makefile:71: test] Error 8</pre>
</details>
</details>

flokli committed:
or-tools: downgrade highs to 1.10.0 again

This is a local revert of 58daf511687f191829238fc7f571e08dc9dedf56,
working around https://github.com/google/or-tools/issues/4911.

flokli commented:
Rebased, and also included https://github.com/NixOS/nixpkgs/pull/459833 to fix the build. Should be rebased again once that PR is merged.

flokli committed:
or-tools: use ' for overrides

This seems to be more consistent with other places in nixpkgs.

GaetanLepage reviewed (approved):
Diff LGTM overall. 
Cleaning the git history would be nice. I think the 3 original commits were the right cut.

Ideally, I would appreciate an additional review from @SomeoneSerge if he has the time and energy for it.

SomeoneSerge commented on pkgs/by-name/or/or-tools/pybind11-2.13.6.nix:
```diff
@@ -0,0 +1,131 @@
+  setupHook = makeSetupHook {
+    name = "pybind11-setup-hook";
+    substitutions = {
+      out = placeholder "out";
```
Unused, I think

flokli replied:
This is simply an old version from nixpkgs without any changes. We can do another round of cleanups in a followup.

SomeoneSerge commented on pkgs/by-name/or/or-tools/scip.patch:
```diff
@@ -0,0 +1,145 @@
+-set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/bin)
+-set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/lib)
+-set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/lib)
++if(CMAKE_PROJECT_NAME EQUAL "SCIP")
```
I don't see this lists included from elsewhere, surprised there would be any other branch...

EDIT: ok looking at targets below I guess it actually is imported...

flokli commented on pkgs/by-name/ab/abseil-cpp_202505/package.nix:
```diff
@@ -0,0 +1,45 @@
+    changelog = "https://github.com/abseil/abseil-cpp/releases/tag/${finalAttrs.version}";
+    license = lib.licenses.asl20;
+    platforms = lib.platforms.all;
+    maintainers = [ lib.maintainers.GaetanLepage ];
```
Yes, this is a plain revert of the package that was already in nixpkgs.
